src/Collection/CollectionTrait.php

```diff
@@ -193,6 +193,54 @@ public function min($callback, $type = SORT_NUMERIC)
     /**
      * {@inheritDoc}
      */
+    public function avg($matcher = null)
+    {
+        $iterator = $this->unwrap();
+        $count = $iterator instanceof Countable ?
+            count($iterator) :
+            iterator_count($iterator);
+
+        if ($count === 0) {
+            return null;
+        }
+
+        return $this->sumOf($matcher) / $count;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function median($matcher = null)
+    {
+        $iterator = $this->unwrap();
+        $count = $iterator instanceof Countable ?
+            count($iterator) :
+            iterator_count($iterator);
+
+        if ($count === 0) {
+            return null;
+        }
+
+        $middle = (int)($count / 2);
+        $elements = $this;
+        if ($matcher != null) {
+            $elements = $elements->extract($matcher);
+        }
+        $values = $elements->toList();
+        sort($values);
```

Comment:
Better to do the count after sorting, so you would be counting array elements instead of an iterator, which is faster, and also will prevent throwing exceptions when using generators.